### ISSUE TITLE
eamxx: `PIO: FATAL ERROR: File create mode is inconsistent among processes` when partial restart files exist

### ISSUE THREAD

**Issue by ndkeen** [E3SM-Project/E3SM#6843]
I was trying to run some ne1024 benchmark cases on pm-gpu with 512 nodes. Run 3 days with output, write restart on last day. Case completed OK once, but the timing was much slower than expected, so I simply resubmitted. Again the timing was slow (it's ongoing issue, but not the point here) and resubmitted. However, the next 4 jobs all failed with following error:

```
 908: PIO: FATAL ERROR: Aborting... FATAL ERROR: File create mode is inconsistent among processes. (file = cess-v2-cntl.ne1024pg2_ne1024pg2.F2010-SCREAMv1.se18-oct18.n0512t4x111XX1.allyamlhist.3d.wr.sk.dins.cice.r.2019-08-04-00000.nc) (/dvs_ro/cfs/cdirs/\
e3sm/ndk/repos/se18-oct18/externals/scorpio/src/clib/pioc_support.c: 3418)

/pscratch/sd/n/ndk/e3sm_scratch/pm-gpu/se18-oct18/cess-v2-cntl.ne1024pg2_ne1024pg2.F2010-SCREAMv1.se18-oct18.n0512t4x111XX1.allyamlhist.3d.wr.sk.dins
```

@mahf708 said he also hit same error (I think also with ne1024 cases on pm-gpu) and found that he needed to remove any partial restart files -- perhaps `cice.r` in particular.

Now in my case, I'm *not* trying to read from restarts -- only writing. But for the fails, there were some partial restarts.
I've already removed the restarts (and rpointers) that were there as I'm testing if next job will be ok.

Labelling as eamxx, but not sure if it's specific to eamxx.

**Comment by ndkeen** [comment]
OK I simply removed all of the partial restarts and resubmitted.  Job was ok.  This suggests to me, that the code is perhaps looking at existing files before trying to open and write. Which is maybe not what we want.

**Comment by ndkeen** [comment]
Happened again as I forgot about this.

```
 544: PIO: FATAL ERROR: Aborting... FATAL ERROR: File create mode is inconsistent among processes. (file = cess-v2-cntl.ne1024pg2_ne1024pg2.F2010-SCREAMv1.c07-mar5.n0512t4x111XX1.allyamlhist.3d.wr.sk.dins.cice.r.2019-08-04-00000.nc) (/dvs_ro/cfs/cdirs/e3sm/ndk/repos/c\
07-mar5/externals/scorpio/src/clib/pioc_support.c: 3420)
```

**Comment by ndkeen** [comment]
Ran into this issue again with ne256 case on pm-gpu.  It's still not obvious how it's happening, but I think it's clear which file to remove and try again.

**Comment by bartgol** [comment]
So the error happens with a CICE file...but does it only happen with EAMxx compsets? That'd be weird...

**Comment by ndkeen** [comment]
Yes all of these are with eamxx compsets

**Comment by ndkeen** [comment]
Happened again, but I'm pretty sure there was not an existing restart file. Will at least record which file it's writing out. Here it's a `.h` file.

```
440: PIO: FATAL ERROR: Aborting... FATAL ERROR: File create mode is inconsistent among processes. (file = u.ne256pg2_ne256pg2.F2010-SCREAMv1.c10-mar26.n0128.spanc860.cice.h.2020-02.nc)
```

**Comment by AaronDonahue** [comment]
@dqwu have you had a chance to look at this?

**Comment by dqwu** [comment]
> [@dqwu](https://github.com/dqwu) have you had a chance to look at this?

Actually we discussed a similar issue in April 2021 with @whannah1(reproduced on Summit)
`PIO: FATAL ERROR: Aborting... FATAL ERROR: File create mode is inconsistent among processes. (/autofs/nccs-svm1_home1/hannah6/E3SM/E3SM_SRC2/externals/scorpio/src/clib/pioc_support.c: 2635)`

@jayeshkrishna had some comments at that time:

> This error typically happens due to application logic that results in different flags passed to Scorpio while creating a file (rewrite the file vs create new file etc)
> 
> the application logic might have race conditions that is not handled correctly resulting in different flags being passed while creating the file, ideally the application should synchronize the approach across all procs

@AaronDonahue I am not familiar with EAMxx code, do you know where pio_createfile is called? We can definitely add some MPI calls and debug prints there to debug this issue.

**Comment by AaronDonahue** [comment]
@dqwu , yes.  The direct call to `PIOc_createfile` is here: https://github.com/E3SM-Project/E3SM/blob/master/components/eamxx/src/share/io/eamxx_scorpio_interface.cpp#L424

This function is called by the `register_file` routine here: https://github.com/E3SM-Project/E3SM/blob/master/components/eamxx/src/share/io/eamxx_output_manager.cpp#L200

Note that line above is for the restart_history files.  You can search elsewhere in `eamxx_output_manager.cpp` for other instances of `register_file` being called.

**Comment by rljacob** [comment]
This is a problem in CICE code, not Eamxx code.  Its an EAMxx problem in the sense that you guys are the only ones keeping CICE alive for your high res F cases.

**Comment by AaronDonahue** [comment]
Any reason we haven't switched entirely over to MPASS-I for EAMxx runs?  @crterai ?

**Comment by rljacob** [comment]
There were scaling and performance issues which I don't think have been addressed.

**Comment by dqwu** [comment]
In Jan 2024, @aldivi reported a similar issue on Compy:
`so ice_pio_open() in the ice log has the correct mode values: 512 when a file exists and 516 when a file does not exist. But the same diagnostic output in e3sm log is mixed (512 or 516) as is the diagnostic output from PIOc_createfile(). So it appears that if no cice history files exist then it may work ok, but if cice history files do exist it breaks, but inconsistently. This time it went a few more months ok, then failed again.`

In components/cice/src/drivers/cpl/ice_pio.F90, we have code like the following:
```
inquire(file=trim(filename), exist=exists)
if (exists) then
   ...
   nmode = pio_clobber
   ...
   status = pio_createfile(ice_pio_subsystem, File, pio_iotype, trim(filename), nmode)
   ...
else
   nmode = pio_noclobber
   ...
   status = pio_createfile(ice_pio_subsystem, File, pio_iotype, trim(filename), nmode)
   ...
endif
```
This approach is not reliable, since the inquire statement may return different logical values for exists on different processors.

In fact, in many places in the E3SM codebase, the inquire(file=...) call is made only on the master processor.

@jayeshkrishna gave a reasonable solution to @aldivi , but unfortunately that fix was never applied:

> Try to reduce the result of the inquire() call (inquire(file=trim(filename),exist=exists)) and use it to decide on the flags to use for creating the file. The decision needs to be consistent across all procs (e.g. Some procs should not use exists = true while others use exists = false : which can potentially result in mismatch on the clobber flag among the procs)
> (Or just do the inquire check on proc 0 and bcast the result to other procs)

**Comment by dqwu** [comment]
As suggested by @jayeshkrishna, we can just add "if (masterproc)" around the inquire call and then broadcast the exists value to other procs. A PR will come soon.
